### ISSUE TITLE
generalize file_handlers to handle BVPs, EVPs

### ISSUE THREAD

**Issue by bpbrown** [DedalusProject/dedalus#237]
It would be very helpful if we could use `FileHanders` to cleanly create hdf5 output files for `BVP` and `EVP` problem types.  This would let us write solutions to disk in a clean way that could be used for other tasks (e.g., `BVP` solutions that are read from disk to initialize an `IVP` or for `EVP`s).

Right now, `FileHandlers` appears to assume that `IVP` related metadata is present in the solver (e.g., `sim_time`).  If, after completing a `NLBVP` solution one tries to write the solution to disk using a filehandler like so:
```
solution = solver.evaluator.add_file_handler(data_dir+'/'+case_dir, mode='overwrite')
solution.add_task(b)
solution.process()
```
then the following kind of error is thrown:
```
Traceback (most recent call last):
  File "unsaturated_atm.py", line 158, in <module>
    solution.process()
  File "/Users/bpbrown/conda_install/src/dedalus-d3/dedalus/core/evaluator.py", line 566, in process
    self.write_file_metadata(file, write_number=self.total_write_num, **kw)
  File "/Users/bpbrown/conda_install/src/dedalus-d3/dedalus/core/evaluator.py", line 619, in write_file_metadat
    super().write_file_metadata(file, **kw
  File "/Users/bpbrown/conda_install/src/dedalus-d3/dedalus/core/evaluator.py", line 585, in write_file_metadata
    dset[self.file_write_num-1] = kw[name]
KeyError: 'sim_time'
```

A workaround right now is to explicitly trigger an evaluation via `Evaluator`, with the various IVP metadata explicitly set to zero:
```
solution = solver.evaluator.add_file_handler(data_dir+'/'+case_dir, mode='overwrite')
solution.add_task(b)
solver.evaluator.evaluate_handlers([solution], timestep=0, wall_time=0, sim_time=0, iteration=0, world_time=0)
```
This works and writes files to disk as expected.

It would be nice to have a more graceful solution, which doesn't require the dummy meta-data, and includes whatever metadata we think is relevant to a BVP or EVP instead.

**Comment by kburns** [comment]
For BVPs it seems like just a save method for individual fields might be the simplest way to save the final solutions. These will generally not be distributed, so I'm not sure the FileHandler machinery adds much.  Maybe if we wanted to save every iteration it could be useful to organize the sequential writes, though.

For EVPs things are pretty different -- the data isn't really field data, it's whole arrays of eigenvalues and eigenvectors.  This would probably not be very elegant to jam into the FileHandler framework -- probably just standalone save/load methods for the EVP solver would make more sense?

**Comment by bpbrown** [comment]
I think it's probably useful to have things grouped in a single file for the BVPs too; what if you want to save the entire state from a hydrostatic/thermal equilibrium solve in an atmosphere, for later use in something like EVPs or IVPs?  There, doing field-by-field saves might be less elegant than "here's a single file that has the whole system".  But maybe that's too special cased to be useful to design for.

Good questions.

**Comment by kburns** [comment]
Working on this in #238

**Comment by kburns** [comment]
Closed with #238.
